### ISSUE TITLE
Not receiving Data after WS2 reconnection

### ISSUE THREAD

**Issue by s5m4** [bitfinexcom/bitfinex-api-node#233]
Hi,
I realized the websocket connection just breaks off after some hours and doesn't reconnect although autoReconnect is set to true. Now I am trying to get a manual reconnection by first connecting the websocket and subscribing to some candle data. Then the websocket.close() function gets called to simulate a connection error at runtime. In the on close function I set another timeout and try to create a new BFX() object and open() it but .on(open) never gets called. The following code works, just copy paste it and run to see for yourself. I am very thankful for any hints or tips.  

```javascript
const BFX = require('bitfinex-api-node');

//websocket
const opts = {
  manageCandles: true, 
  transform: true,
  API_KEY: 'hidden',
  API_SECRET: 'hidden',
  ws: {
    autoReconnect: true,
    seqAudit: true,
    packetWDDelay: 10 * 1000
  }
};
var websocket = new BFX().ws(2, opts);

websocket.on('open', () => {
	//websocket.auth.bind(websocket)
	console.log('.on(open) called');
	websocket.subscribeCandles('trade:5m:tBTCUSD')
});

websocket.on('close', ()=>{
	console.log('.on(close) called');
	setTimeout(function() { 
		websocket = new BFX().ws(2, opts);
		websocket.open();
	}, 10000);
});

websocket.onCandle({key: 'trade:5m:tBTCUSD'},(candles) => {	
	console.log(candles[0]);
})

websocket.open();

// this closes the connection after 20 seconds 
//after start for debugging purposes:
setTimeout(function() { 		
		websocket.close();
	}, 10000);
```

**Comment by s5m4** [comment]
#138 This might be related.

**Comment by f3rno** [comment]
@s5m4 don't reconnect by making a new `BFX` object; there is a `reconnect()` function on `WSv2` :) So you can do:

```js
websocket = new BFX().ws(2, opts);

// later...
websocket.reconnect().then(() => ...).catch(err => ...)
```
If it still doesn't work properly after using that method, let me know and I will take a closer look. If you make two BFX instances like that, you'll create 2 ws2 connections (one might be hanging)

**Comment by s5m4** [comment]
@f3rno thanks for your reply! I wonder why it didn't strike me to look into the websocket specs to get the .reconnect() function. The webserver is reconnecting now it seems ( on.open() gets called) but there are no events triggering the .on(candle) function. I even modified the code to avoid calling the .subscribeCandles() twice because I thought that maybe subscribing a second time acts as a unsubscribe but nope... 
Here's what I have now. If you could spend a few minutes to try yourself I'd be more than thankful!

```javascript
var BFX = require('bitfinex-api-node');

//websocket
const opts = {
  manageCandles: true, 
  transform: true,
  API_KEY: 'hidden',
  API_SECRET: 'hidden',
  ws: {
    autoReconnect: false,
    seqAudit: false,
    packetWDDelay: 10 * 1000
  }
};

var websocket = new BFX().ws(2, opts);
var reconnect = false;

websocket.on('open', () => {
	//websocket.auth.bind(websocket)
	console.log('.on(open) called');
	if(!reconnect){
		websocket.subscribeCandles('trade:5m:tBTCUSD')
	}
});

websocket.on('close', ()=>{
	console.log('.on(close) called');
	
	setTimeout(function() { 
		websocket.reconnect().then(reconnect = true).catch((err) => {console.log(err);}) 
	}, 10000);
});

websocket.onCandle({key: 'trade:5m:tBTCUSD'},(candles) => {	
	console.log(candles[0]);
})

websocket.open();

// this closes the connection after 10 seconds 
//after start for debugging purposes:
setTimeout(function() { 		
		websocket.close();
	}, 10000);

```

**Comment by f3rno** [comment]
@s5m4 I'm looking at this now, should have a solution over today

**Comment by mescalina** [comment]
Hi,

the connection gets open and closed continuosly after the reconnect because of this option: packetWDDelay: 10 * 1000

It should forcing the connection to keep the connection alive but it results in multiple open events being fired.

I have the same problem with trades, after successfully reconnecting to tha api listeners are not working anymore.

**Comment by jinxmcg** [comment]
soo...any update? I have the same issue

**Comment by mescalina** [comment]
No news no new regrets..
I ended up writing my own bitfinex module and it usually successfully reconnect destroying the old connection and listeners first and recreating new ones. But still 1 time during a 1 week period it reconnected but listeners did't work. This problem is simply horrible because it makes the module totally unreliable on the most important feature which is staying connected.

**Comment by jsappme** [comment]
I have the same issue. No progress on this one?

**Comment by jsappme** [comment]
@mescalina would you mind sharing your module?

**Comment by f3rno** [comment]
@mescalina @jsappme apologies for the delay guys, I got swamped with other work. I'm working on getting the next beta release out ASAP, taking review time into account it should be released next week at the latest.

**Comment by robertkowalski** [comment]
I am sorry for the delays and that you felt alone. In understand that it frustrates when something is not working properly. Still want you to remember: this project is maintained by humans and sometimes humans have to work on other things. 

The bfx api client is an open source project, **nobody has to use it**. Everyone participating, from core committers to users can and is encouraged to step up and work together with us. There is also always the option to revert to later versions.

Open source lives from constructive feedback and an environment were people work and thrive together. 

Be a positive helper, don't be a negative complainer!

**Comment by jsappme** [comment]
@f3rno @robertkowalski if I can help with anything let me know. Thanks for your work.

**Comment by f3rno** [comment]
@mescalina @jsappme I've fixed a sporadic reconnect issue in `2.0.0-beta.1`/`master`, if this is still occurring for you please update your `bitfinex-api-node` and test again. If it happens afterwards, re-open this issue :slightly_smiling_face:

**Comment by zenymax** [comment]
@f3rno it's occurring with me. I 'm testing on the newest version. Can you teach me what do you do to fix it? thanks very much for hard work.
